velox/type/TimestampConversion.cpp

```diff
@@ -826,6 +840,158 @@ Status parserError(const char* str, size_t len) {
       std::string(str, len));
 }
 
+// Spark supports an id with one of the prefixes UTC+, UTC-, GMT+, GMT-, UT+ or
+// UT-, and a suffix in the following formats: h[h], hh[:]mm, hh:mm:ss, hhmmss.
+// This function will check if the suffix format is correct. Furthermore, this
+// function adds the seconds offset to the input timestamp because
+// IANA(https://www.iana.org/time-zones) does not support seconds in the offset.
+bool parseSparkTzWithPrefix(std::string& timeZoneName, Timestamp& ts) {
+  std::string_view suffix{timeZoneName};
+  if (suffix.length() < 2) {
+    return false;
+  }
+  auto sign = timeZoneName[0];
+  std::string_view offset = suffix.substr(1);
+  std::string_view hm, s;
+  bool appendZeroMinutes = false;
+  if ((offset.length() == 1 && characterIsDigit(offset[0])) ||
+      (offset.length() == 2 && allCharactersIsDigit(offset))) {
+    // h[h]
+    hm = offset;
+    // Use appendZeroMinutes to indicate that we need to append :00 to the end
+    // of timeZoneName. By appending only when assigning to timeZoneName, hm can
+    // be defined as a string_view to minimize multiple string copies.
+    appendZeroMinutes = true;
+  } else if (
+      (offset.length() == 4 && allCharactersIsDigit(offset)) ||
+      (offset.length() == 5 && allCharactersIsDigit(offset, 0, 1) &&
+       allCharactersIsDigit(offset, 3, 4))) {
+    // hh[:]mm
+    // TimeZoneMap::normalizeTimeZoneOffset will normalize hhmm to hh:mm
+    hm = offset;
+  } else if (
+      (offset.length() == 6 && allCharactersIsDigit(offset)) ||
+      (offset.length() == 8 && allCharactersIsDigit(offset, 0, 1) &&
+       allCharactersIsDigit(offset, 3, 4) &&
+       allCharactersIsDigit(offset, 6, 7))) {
+    // hh:mm:ss, hhmmss
+    auto lastSep = offset.rfind(':');
+    if (lastSep == std::string::npos && offset.length() == 6) {
+      // hhmmss
+      // TimeZoneMap::normalizeTimeZoneOffset will normalize hhmm to hh:mm.
+      hm = offset.substr(0, 4);
+      s = offset.substr(4);
+    } else if (lastSep == 5 && offset.find(':') == 2 && offset.length() == 8) {
+      // hh:mm:ss
+      hm = offset.substr(0, 5);
+      s = offset.substr(6);
+    } else {
+      return false;
+    }
+  } else {
+    return false;
+  }
+  int32_t sec = 0;
+  size_t pos = 0;
+  if (!s.empty() && !parseDoubleDigit(s.data(), s.length(), pos, sec)) {
+    return false;
+  }
+  if (sec < 0 || sec > 60) {
+    return false;
+  }
+  if (sec != 0) {
+    // Seconds timezone offset is not supported,
+    // and needs to be applied to timestamp.
+    if (sign == '+') {
+      if (ts.getSeconds() - sec < Timestamp::kMinSeconds) {
+        return false;
+      }
+      ts = Timestamp(ts.getSeconds() - sec, ts.getNanos());
```

Comment:
There is a uniform adjustment by `ts.toGMT(timezone)` based on the returned timestamp and timezone from `fromTimestampWithTimezoneString`.

Can we somehow let `ts.toGMT(timezone)` do your proposed adjustment based on `sec`?

Reply:
@liujiayi771, I get your point. I mean can we extract the adjustment based on `sec` and somehow make `ts.toGMT` include this adjustment? Then, here we can only do timezone normalization. This decouple may be helpful to support normalizing timezone from config.

Another question is, do we really need to support timezone with second-level offset? It is very rare to use such timezone in real world. Actually almost all timezones have hour, half hour or quarter hour level offset. If there is no real request, maybe we can just ignore such timezone in the implementation, even though Spark supports it.

Reply:
@PHILO-HE I see, I can extract this logical to a method in `Timestamp`.

Reply:
@liujiayi771, then, I recommend to ignore such timezone with second-level offset. You can ask for others' feedback in the community. Thanks!

Reply:
> +|-h[h]:mm:ss
+|-hhmmss

@liujiayi771 I notice you mentioned above two kinds of suffix for the timezone id, and want to know where they come from. Does Spark mention such requirement? Thanks.

Reply:
@rui-mo They come from the comments in Spark code.
https://github.com/apache/spark/blob/d3d84e045cc484cf7b70d36410a554238d7aef0e/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala#L288-L289
`+|-h:mm:ss` is not present in the comments, but I have tested this pattern can work.